### PR TITLE
Django DB to GCP Pipeline Application Commit (iss. #1, #2, #7)

### LINKED ISSUE #1: application configuration
Some things about this application should be configurable, especially for deployment to OpenShift.  For example…

* DB credentials
* GCP credentials
* number of months to look back for data

These things will be added to the project in OpenShift as a secret or configmap.  That way, they do not need to be committed to this repo in order to make them available on OpenShift.  This is a very good way to prevent credentials from becoming public.

Some configuration items should have default values, like the number of months.  Credentials definitely should not have defaults and an error should be given if the credentials aren't available.

### LINKED ISSUE #2: documentation (`README.md`, etc.)
The project has rough documentation in the `steps.md` file.  Make new documentation in the file `README.md`.  It can be a cleaned-up form of the current docs and there's probably more to be added.  If the documentation seems to be too long for a single file, it can be broken into other files, with links from the README.

### LINKED ISSUE #7: use logging library
_Originally posted by @lsloan in https://github.com/tl-its-umich-edu/mpr-research-data/pull/3#discussion_r868451198_
> That sounds like a good solution.  I think it effectively disables buffering.
>
> We may want to switch to some sort of standard logging, but we can leave that to another PR.  I'll make an issue for that, too.

Replace printing of various status and error messages with calls to the Python logging library.  That will put messages into a standard format, including timestamp, message type (info, error, debug, etc.), and the line in the code where the logging occurred.

### PR DESCRIPTION
**Pull request by takposha**
Currently working/Completed:
dbToBucketScript.py can access the DB, retrieve Course IDs, then retrieve corresponding Course data and send them to a GCP bucket.
Error messages and comments have been applied across the script so it should be easier to know if and why something does not work.
The SQL queries can be modified using variables the user provides.
There is a basic README file now. It needs to be updated for config parameters.

Not working/needs to be done:
I'm not sure how a config file should be made for a Docker application. I have listed all variables that are to be config ones in the python script files, so it should just be a matter of moving them into a config file. Will use a .env file setup to do this.
I don't know how frequent data calls are, and if the files get updated, but maybe having some way to check if the GCP bucket TSV file already exists and is up to date can help avoid unnecessary data pushes to the bucket. This shouldn't matter if it's only a small amount of data daily.

Resolves #1 
Resolves #2 
Resolves #7

### CODE REVIEW

**Comment by takposha** [comment]
Added a config file for adjusting variables. Updated ReadMe for steps on config file management. 
Adjust Docker files to read .env file and disabled sleep infinity.

**Review by lsloan — CHANGES_REQUESTED** [review]
My apologies to keep you waiting for the review.  Please don't be alarmed over the number of comments I gave.  A lot of what I mention is about standard practices.  They're things you probably weren't aware of and you're likely to avoid them in your future commits.

You will probably avoid some problems if you follow Python's [PEP 8](https://peps.python.org/pep-0008/) document.  I agree with most of it.  (I like `camelCase` variable names instead of their recommended `underscored_words`.)  There are programs that can analyze your code and make sure it meets the standards.  I'm considering adding that as a check on the GH repo.  We've used it for other projects.

I've given enough comments for now, I think.  Work through some of them and we'll do another round of code review.  On the next round, if there's anything significant outstanding, I'll open new issues to resolve them.

File: README.md
Comment (takposha) [reply README.md]: Done.

File: .gitignore
Comment (takposha) [reply .gitignore]: Will address in next comment.

File: README.md
Comment (takposha) [reply README.md]: Noted for future.

File: README.md
Comment (takposha) [reply README.md]: Noted, Fixed.

File: SQL Query Files/courseQuery.sql
Comment (takposha) [reply SQL Query Files/courseQuery.sql]: So, regarding queries. I have placed them in a simpler named folder. I will maybe have a way to instead allow for default queries to be called? This would also need some work to catch custom query formatting, like the number of months and such. 

The simplest idea then would just paste a whole query in the .env that is run as-is.

File: dbToBucketLibrary.py
Comment (takposha) [reply dbToBucketLibrary.py]: Done everywhere where was needed.

File: config.env
Comment (takposha) [reply config.env]: I might need to look more into the GCloud Auth methods, as the Auth methods right now want an actual JSON file path. 
I make a JSON key file temporarily and immediately destroy it after the connection is made, but this feels a bit janky and I would like to fix that.

File: dbToBucketScript.py
Comment (takposha) [reply dbToBucketScript.py]: I put it all into one file. This is a habit of mine running code in Jupyter where I didn't care about the functions, and mostly wanted to look at the outputs, so I'd put everything in a separate file.

Comment (takposha) [reply dbToBucketScript.py]: I agree with these points. To be fair, I never thought about naming for anyone besides me, so that's good to keep in mind.

File: dbToBucketScript.py
Comment (takposha) [reply dbToBucketScript.py]: I think I found a fix, using a variable PYTHONUNBUFFERED set to 1 in the dockerfile. It should do the same thing as flush.

File: start.sh
Comment (takposha) [reply start.sh]: Done, added that to start.sh.

**Comment by takposha** [comment]
That should be the last commit. VSCode PEP 8 formatting doesn't catch everything PepSpeaks does. But that should be sorted now.

File: access data bucket.ipynb
Comment (lsloan) [reply access data bucket.ipynb]: Thanks!

File: SQL Query Files/courseQuery.sql
Comment (lsloan) [reply SQL Query Files/courseQuery.sql]: I found one file that needed a newline (`queries/courseQuery.sql`), but I added it myself, so you don't need to worry about it.

File: README.md
Comment (lsloan) [reply README.md]: Thanks!

File: README.md
Comment (lsloan) [reply README.md]: Thanks!

File: SQL Query Files/courseQuery.sql
Comment (lsloan) [reply SQL Query Files/courseQuery.sql]: I think we can leave the concept of configurable queries for another issue in the next few days.  We can work with the queries in separate files for now.  You raise good questions about handling custom and default queries.  I'll try to make suggestions in a new issue.

Comment (lsloan) [reply SQL Query Files/courseQuery.sql]: Thanks!

File: config.env
Comment (lsloan) [reply config.env]: ✅

File: config.env
Comment (lsloan) [reply config.env]: ✅

File: config.env
Comment (lsloan) [reply config.env]: ✅

File: config.env
Comment (lsloan) [reply config.env]: Oh, it requires a file?  If you can't find a way around that, then making the key file from the application is not such a bad thing.  There are Python library functions for creating temporary files, like in `/tmp` on Linux environments.  Since nothing other than our application will ever access that environment, I think it's safe to create the file if it doesn't already exist and leave it there for the next time the application runs.

File: config.env
Comment (lsloan) [reply config.env]: ✅

File: docker-compose.yml
Comment (lsloan) [reply docker-compose.yml]: ✅

File: dbToBucketScript.py
Comment (lsloan) [reply dbToBucketScript.py]: Oh, sure.  I didn't think of that.  I've done that in Jupyter myself.

Comment (lsloan) [reply dbToBucketScript.py]: Thanks!

File: dbToBucketLibrary.py
Comment (lsloan) [reply dbToBucketLibrary.py]: ✅

File: dbToBucketLibrary.py
Comment (lsloan) [reply dbToBucketLibrary.py]: ✅

File: dbToBucketLibrary.py
Comment (lsloan) [reply dbToBucketLibrary.py]: ✅

File: dbToBucketScript.py
Comment (lsloan) [reply dbToBucketScript.py]: ✅

File: dbToBucketScript.py
Comment (lsloan) [reply dbToBucketScript.py]: That sounds like a good solution.  I think it effectively disables buffering.

We may want to switch to some sort of standard logging, but we can leave that to another PR.  I'll make an issue for that, too.

File: dbToBucketScript.py
Comment (lsloan) [reply dbToBucketScript.py]: ✅

File: dbToBucketScript.py
Comment (lsloan) [reply dbToBucketScript.py]: ✅

**Comment by lsloan** [comment]
@pep8speaks suggest diff

File: mpr-research-data.py
Comment (lsloan) [inline mpr-research-data.py:83]: Try `from_service_account_info()`…


```suggestion
        client = storage.Client.from_service_account_info(gcpJSON)
```

From docs (hopefully correct ones) at: https://cloud.google.com/python/docs/reference/aiplatform/latest/google.cloud.aiplatform_v1beta1.services.model_service.ModelServiceClient

File: dbToBucketScript.py
Comment (lsloan) [reply dbToBucketScript.py]: Opened #7.

File: start.sh
Comment (lsloan) [reply start.sh]: I didn't know about that `--iso-8601=seconds` option.  I used something different.  I'll check that out.

File: config.env
Comment (lsloan) [reply config.env]: I added a suggestion that I think may work.  See comment below (https://github.com/tl-its-umich-edu/mpr-research-data/pull/3#discussion_r868788736).

File: dbToBucketLibrary.py
Comment (lsloan) [reply dbToBucketLibrary.py]: ✅

File: .gitignore
Comment (lsloan) [reply .gitignore]: ✅

File: mpr-research-data.py
Comment (takposha) [reply mpr-research-data.py:83]: I have implemented that. I also made the .env file consistent in how the two key values are taken.

**Comment by takposha** [comment]
Added the better GCP key implementation.

Added logging to console using the logging library.

File: mpr-research-data.py
Comment (lsloan) [inline mpr-research-data.py:143]: I suggest naming variables a little more specifically, if they're known to always contain specific data.

Also, although the SQL query is nicely formatted, the parameter substitution doesn't need to be neat also.  Processing the substitution can be a little simpler.

```suggestion
def retrieveQueryMaker(retrieveQueryTemplate, courseIDs, engine):

    try:
        courseIDString = ','.join(map(str, courseIDs))
```

**Review by lsloan — CHANGES_REQUESTED** [review]
I suggest another change to one query handler.  Handle that, then I think I can merge this PR.

For the most part, we like issues to have about a 1:1 correspondence with PRs.  We don't add more issue resolutions to one PR because the PR will stay open for a long time and the conversations overlap and get confusing.

The correspondence doesn't need to be strictly 1:1.  PRs can address issues that seem to be closely related.  For this project, because it's small and there aren't many issues yet, we'll try to stick to 1:1.

It's helpful to relate issues to their PRs.  Do that by adding text to the PR description, like one of the following…

```md
Resolves #𝒏
Fixes #𝒏
Closes #𝒏
```

In this case, you've added resolutions for a few issues, so mention each of them in the description.

We also find it useful to mention all the resolved issues in the PR title, too.  That is, add to the end of it something like, `(Iss. #𝒏, #𝒏, #𝒏, …)`.  That makes it easy for us to see the relationships when we're looking at a list of issues and PRs.

**Comment by lsloan** [comment]
I thought what you had before with the GCP key JSON embedded in the `.env` file was a good thing.  I think of that key in JSON as a single entity.  We don't know what all of its contents will be.  Google may change it at some point.  So, if we break it down into its components now, maybe in the future it won't work.  It seems better to keep it as a single JSON string.

My change reintroduced the json module, so I organized the imports, too.  It's fairly common to organize imports by core Python modules, followed by third-party modules, and lastly local project modules.

**Comment by lsloan** [comment]
So, the good news is that when I ran the application that I checked out of your branch, it worked well.

Aside from the GCP key change I made, I made a couple of changes to my `.env` file, too.  I changed `NUMBER_OF_MONTHS = 1` to run a shorter test.  And I changed `GCLOUD_BUCKET = 'mpr-research-data-uploads-lsloan_test'`, to write to a different bucket for my test, so it wouldn't disturb what was already there.

I ran the app before I created the new bucket, just to see what would happen.  I got results like this:

```txt
mpr-research-data  | 2022-05-10T14:44:52+0000 INFO     [mpr-research-data.py:170] - Slicing: 495677 - Math 216 WN 2022.tsv
mpr-research-data  | 2022-05-10T14:44:52+0000 INFO     [mpr-research-data.py:181] - Saving to GCP: 495677 - Math 216 WN 2022.tsv
mpr-research-data  | 2022-05-10T14:44:56+0000 ERROR    [mpr-research-data.py:186] - Error Message: 404 POST https://storage.googleapis.com/upload/storage/v1/b/mpr-research-data-uploads-lsloan_test/o?uploadType=multipart: {
mpr-research-data  |   "error": {
mpr-research-data  |     "code": 404,
mpr-research-data  |     "message": "The specified bucket does not exist.",
mpr-research-data  |     "errors": [
mpr-research-data  |       {
mpr-research-data  |         "message": "The specified bucket does not exist.",
mpr-research-data  |         "domain": "global",
mpr-research-data  |         "reason": "notFound"
mpr-research-data  |       }
mpr-research-data  |     ]
mpr-research-data  |   }
mpr-research-data  | }
mpr-research-data  | : ('Request failed with status code', 404, 'Expected one of', <HTTPStatus.OK: 200>)
mpr-research-data  | 2022-05-10T14:44:56+0000 ERROR    [mpr-research-data.py:187] - Failed to upload Course Data for 495677 - Math 216 WN 2022.tsv to GCP.
```

Which is good.  It reported that once for each of the courses that met my 1-month criteria.

So, I created the required bucket, ran it again, and all was good.

**Review by lsloan — APPROVED** [review]

**Pull request merged**
